File: server/api/stats.py
import time
import math

import numpy as np
from tornado import escape, web

from distribution import Distribution
from grids.spiral import SpiralGrid, SPIRAL_GRID_POINTS
from grids.classic import ClassicGrid, CLASSIC_GRID_DIV


class StatsAPI(web.RequestHandler):
    def initialize(self, database):
        self.database = database

    def __getArgs(self, body):
        data = escape.json_decode(body)

        self.points = data.get("points")
        self.mean = data.get("mean")
        self.cov = data.get("cov")

    def __validate(self):
        if self.mean is None or self.cov is None or self.points is None:
            raise ValueError("not enough arguments")

        self.points = np.reshape([float(point) for point in self.points.split(',')], (-1, 3))
        self.mean = tuple([float(coord) for coord in self.mean.split(',')])
        #length = np.sqrt(np.sum([x * x for x in self.mean]))
        #self.mean = tuple(x / length for x in self.mean)
        self.cov = tuple(float(cov) for cov in self.cov.split(','))
        # TODO len check

    def __getGrids(self):
        key = (self.mean, self.cov)

        if not (key in self.database):
            distribution = Distribution(self.mean, self.cov)

            self.database[key] = (
                SpiralGrid(SPIRAL_GRID_POINTS, distribution.calc),
                ClassicGrid(CLASSIC_GRID_DIV, distribution.calc)
            )

        self.spiral_grid, self.classic_grid = self.database[key]

    def __calculateTStats(self, value):
        return self.spiral_grid.calculateIntegralInsideIsoline(value)

    # TODO classmethod?
    def __distanceBetweenPoints(self, a, b):
        return math.sqrt(np.sum([(a[i] - b[i]) ** 2 for i in range(3)]))

    # TODO move/rename all methods below?
    def __insertPointIntoIsoline(self, point, isoline_points):
        point_index = None

        distances = np.array([self.__distanceBetweenPoints(iso_point, point) for iso_point in isoline_points])
        min_distances_by_indices = distances.argsort()[:2]
        indices_diff = int(math.fabs(min_distances_by_indices[0] - min_distances_by_indices[1]))
        if indices_diff == len(isoline_points) - 1:
            point_index = len(isoline_points)
        elif indices_diff == 1:
            point_index = np.max(min_distances_by_indices)
        else:
            print("WARNING: 2 min distances are not nearby!", isoline_points, distances, min_distances_by_indices, point)
            raise False  # TODO
        isoline_points.insert(point_index, point)  # TODO check it by drawing 3 points explicitly

        return point_index, isoline_points

    def __getNormalizedVector(self, point_a, point_b):
        vector = [point_b[i] - point_a[i] for i in range(len(point_a))]
        vector_len = np.linalg.norm(vector)
        return [x / vector_len for x in vector]

    def __getPrevIndexOfIsolinePoint(self, point_index, isoline_points):
        if point_index == 0:
            return len(isoline_points) - 1
        return point_index - 1

    def __getNextIndexOfIsolinePoint(self, point_index, isoline_points):
        if point_index == len(isoline_points) - 1:
            return 0
        return point_index + 1

    # TODO comments!
    def __checkClockwiseDirection(self, M_vector, I_vector, A_vector):
        AI_vector = np.cross(A_vector, I_vector)
        AI_proj_on_M = np.dot(AI_vector, M_vector)

        if AI_proj_on_M > 0:
            return True
        return False

    # TODO gradient
    def __calcIntegralOnFullIsoline(self, isoline_points):
        integral = np.sum([self.__distanceBetweenPoints(isoline_points[i], isoline_points[i + 1]) for i in range(len(isoline_points) - 1)])
        return integral + self.__distanceBetweenPoints(isoline_points[len(isoline_points) - 1], isoline_points[0])

    def __calcIntegralOnCurveOnIsolineLeft(self, isoline_points, start_index, end_index):
        if end_index < start_index:
            return np.sum([self.__distanceBetweenPoints(isoline_points[i], isoline_points[i + 1]) for i in range(end_index, start_index)])
        
        first_part = np.sum([self.__distanceBetweenPoints(isoline_points[i], isoline_points[i + 1]) for i in range(start_index)])
        second_part = np.sum([self.__distanceBetweenPoints(isoline_points[i], isoline_points[i + 1]) for i in range(end_index, len(isoline_points) - 1)])
        return first_part + second_part + self.__distanceBetweenPoints(isoline_points[len(isoline_points) - 1], isoline_points[0])

    def __calcIntegralOnCurveOnIsolineRight(self, isoline_points, start_index, end_index):
        if start_index < end_index:
            return np.sum([self.__distanceBetweenPoints(isoline_points[i], isoline_points[i + 1]) for i in range(start_index, end_index)])
        
        first_part = np.sum([self.__distanceBetweenPoints(isoline_points[i], isoline_points[i + 1]) for i in range(end_index)])
        second_part = np.sum([self.__distanceBetweenPoints(isoline_points[i], isoline_points[i + 1]) for i in range(start_index, len(isoline_points) - 1)])
        return first_part + second_part + self.__distanceBetweenPoints(isoline_points[len(isoline_points) - 1], isoline_points[0])

    def __calculateSStats(self, point, value):
        isoline_points = self.classic_grid.getIsolineCoords(value)
        point_index, isoline_points = self.__insertPointIntoIsoline(point, isoline_points)

        theta, phi = self.classic_grid.getAnglesOfPoint(point)

        mean_normed = self.__getNormalizedVector([0, 0, 0], self.mean)
        theta_mean, phi_mean = self.classic_grid.getAnglesOfPoint(mean_normed)  # TODO change mean with actual pdf's maximum
        if mean_normed[0] == 0 and mean_normed[1] == 0:  # if mean vector points to one of the poles
            theta_mean = 0

        # TODO add comments to this hard logic
        isoline_points_angles = [self.classic_grid.getAnglesOfPoint(iso_point) for iso_point in isoline_points]  # TODO rename
        indices_inside_zero_phi_mean = [i for i in range(len(isoline_points_angles)) if 0 <= isoline_points_angles[i][1] <= phi_mean]
        indices_inside_zero_pi_minus_phi_mean = [i for i in range(len(isoline_points_angles)) if 0 <= isoline_points_angles[i][1] <= (math.pi - phi_mean)]

        diffs_to_theta_mean = [math.fabs(isoline_points_angles[i][0] - theta_mean) for i in indices_inside_zero_phi_mean]
        diffs_to_theta_plus_pi_mean = [math.fabs(isoline_points_angles[i][0] - ((theta_mean + math.pi) % (2 * math.pi))) for i in indices_inside_zero_pi_minus_phi_mean]

        min_diffs_to_theta_mean = np.array(diffs_to_theta_mean).argsort()
        min_diffs_to_theta_plus_pi_mean = np.array(diffs_to_theta_plus_pi_mean).argsort()

        intersection_points = []
        if len(min_diffs_to_theta_mean) > 0:
            intersection_points.append(isoline_points_angles[indices_inside_zero_phi_mean[min_diffs_to_theta_mean[0]]])
            intersection_points.append(isoline_points_angles[indices_inside_zero_phi_mean[min_diffs_to_theta_mean[1]]])  # TODO what if only 1 point?
        if len(min_diffs_to_theta_plus_pi_mean) > 0:
            if len(intersection_points) > 0:
                if diffs_to_theta_mean[min_diffs_to_theta_mean[0]] > diffs_to_theta_plus_pi_mean[min_diffs_to_theta_plus_pi_mean[0]]:
                    intersection_points[0] = (isoline_points_angles[indices_inside_zero_pi_minus_phi_mean[min_diffs_to_theta_plus_pi_mean[0]]])
                    intersection_points[1] = (isoline_points_angles[indices_inside_zero_pi_minus_phi_mean[min_diffs_to_theta_plus_pi_mean[1]]])
            else:
                intersection_points.append(isoline_points_angles[indices_inside_zero_pi_minus_phi_mean[min_diffs_to_theta_plus_pi_mean[0]]])
                intersection_points.append(isoline_points_angles[indices_inside_zero_pi_minus_phi_mean[min_diffs_to_theta_plus_pi_mean[1]]])

        intersection_point = np.mean(intersection_points, axis=0)
        intersection_point_coords = self.classic_grid.getCoordsOfPoint(intersection_point)

        intersection_point_index, isoline_points = self.__insertPointIntoIsoline(intersection_point_coords, isoline_points)

        A_point = isoline_points[self.__getNextIndexOfIsolinePoint(intersection_point_index, isoline_points)]
        A_vector = self.__getNormalizedVector(mean_normed, A_point)
        I_vector = self.__getNormalizedVector(mean_normed, intersection_point_coords)

        isoline_integral = self.__calcIntegralOnFullIsoline(isoline_points)
        curve_integral = 0
        if self.__checkClockwiseDirection(mean_normed, I_vector, A_vector):
            # TODO check direction
            curve_integral = self.__calcIntegralOnCurveOnIsolineRight(isoline_points, intersection_point_index, point_index)
        else:
            curve_integral = self.__calcIntegralOnCurveOnIsolineLeft(isoline_points, intersection_point_index, point_index)

        #return curve_integral / isoline_integral

        return {"isoline": isoline_points, "points": [intersection_point_coords], "pv": [point, value], "S": curve_integral / isoline_integral}


    def post(self):
        start = time.time()
        self.set_status(200)
        self.set_header("Access-Control-Allow-Origin", "*") # TODO

        try:
            self.__getArgs(self.request.body)
            self.__validate()

            self.__getGrids()
            print("StatsAPI: Grids generation and function evaluation time:", time.time() - start)

            start = time.time()

            t = []
            s = []

            self.pointsValues = {tuple(point) : self.spiral_grid.function(point) for point in self.points}
            for point, value in self.pointsValues.items():
                t.append(self.__calculateTStats(value))
                s.append(self.__calculateSStats(point, value))

            print("StatsAPI: Stats calculation time:", time.time() - start)

        except Exception as e:
            self.finish({"code": 500, "body": str(e)})
            raise e  # TODO remove (debug)
            return

        self.finish({"code": 200, "t": t, "s": s})

    # Something bad here, but it's fast workaround right now (CORS policy)
    def options(self):
        self.set_status(200)
        self.set_header("Access-Control-Allow-Origin", "*")
        self.set_header("Access-Control-Allow-Headers", "*")
        if not self.request.body:
            return

File: server/api/stats.cpp
#include "stats.hpp"

#include <algorithm>
#include <array>
#include <cmath>
#include <exception>
#include <limits>
#include <vector>

#include "types/types.hpp"

StatsAPI::StatsAPI(std::vector<CoordsOfPoint>& points, ClassicGrid* classic_grid, SpiralGrid* spiral_grid)
    : points(points), classic_grid(classic_grid), spiral_grid(spiral_grid) {
}

double StatsAPI::DistanceBetweenPoints(const CoordsOfPoint& point1, const CoordsOfPoint& point2) const {
    double sum = 0.;
    for (int i = 0; i < 3; ++i) {
        double diff = point1[i] - point2[i];
        sum += diff * diff;
    }
    
    return std::sqrt(sum);
}

int StatsAPI::InsertPointIntoIsoline(const CoordsOfPoint& point, IsolineCoords& isoline_points) const {
    int point_index = -1;

    if (isoline_points.size() < 2)
        throw std::logic_error("InsertPointIntoIsoline: isoline_points length is less than 2");

    // Calculate distance between point and each isoline point.
    // The smallest two distances will indicate the place, where the point must
    // be inserted into isoline_points array.

    double first_smallest_distance = DistanceBetweenPoints(isoline_points[0], point);
    double second_smallest_distance = DistanceBetweenPoints(isoline_points[1], point);
    int index_of_first_smallest = 0;
    int index_of_second_smallest = 1;

    if (first_smallest_distance > second_smallest_distance) {
        std::swap(first_smallest_distance, second_smallest_distance);
        std::swap(index_of_first_smallest, index_of_second_smallest);
    }

    for (int i = 2; i < isoline_points.size(); ++i) {
        double distance = DistanceBetweenPoints(isoline_points[i], point);
        if (distance < first_smallest_distance) {
            second_smallest_distance = first_smallest_distance;
            first_smallest_distance = distance;

            index_of_second_smallest = index_of_first_smallest;
            index_of_first_smallest = i;

        } else if (distance < second_smallest_distance) {
            second_smallest_distance = distance;
            index_of_second_smallest = i;
        }
    }

    int indices_diff = std::abs(index_of_first_smallest - index_of_second_smallest);
    if (indices_diff == isoline_points.size() - 1)
        // it is the first and the last points of isoline_points
        point_index = isoline_points.size();
    else if (indices_diff == 1)
        // it is the two adjacent points of isoline_points
        point_index = std::max(index_of_first_smallest, index_of_second_smallest);
    
    if (point_index == -1)
        throw std::logic_error("InsertPointIntoIsoline: the smallest two distances are not nearby");

    isoline_points.insert(isoline_points.begin() + point_index, point);

    return point_index;
}

Vector StatsAPI::GetNormalizedVector(const CoordsOfPoint& point1, const CoordsOfPoint& point2) const {
    Vector vec;
    double squared_sum = 0.;

    for (int i = 0; i < 3; ++i) {
        double value = point2[i] - point1[i];

        vec[i] = value;
        squared_sum += value * value;
    }

    double vec_len = std::sqrt(squared_sum);
    for (int i = 0; i < 3; ++i)
        vec[i] /= vec_len;

    return vec;
}

int StatsAPI::GetIndexOfPrevIsolinePoint(int point_index, const IsolineCoords& isoline_points) const {
    if (point_index == 0)
        return isoline_points.size() - 1;
    return point_index - 1;
}

int StatsAPI::GetIndexOfNextIsolinePoint(int point_index, const IsolineCoords& isoline_points) const {
    if (point_index == isoline_points.size() - 1)
        return 0;
    return point_index + 1;
}

bool StatsAPI::CheckClockwiseDirection(const Vector& M, const Vector& I, const Vector& A) const {
    Vector AI; // a cross product of A and I vectors
    AI[0] = A[1] * I[2] - A[2] * I[1]; 
    AI[1] = A[2] * I[0] - A[0] * I[2]; 
    AI[2] = A[0] * I[1] - A[1] * I[0];

    double AI_projection_on_M = 0; // a dot product of AI and M
    for (int i = 0; i < 3; ++i)
        AI_projection_on_M += AI[i] * M[i];
    
    if (AI_projection_on_M > 0)
        return true;
    return false;
}

double StatsAPI::CalcIntegralOnInfinitesimalCurve(const CoordsOfPoint& point1, const CoordsOfPoint& point2) const {
    CoordsOfPoint center_point;
    for (int i = 0; i < 3; ++i)
        center_point[i] = (point1[i] + point2[i]) / 2;

    double gradient_modulus = 1; // TODO np.linalg.norm(self.grad_of_d(middle_point)) // self.grad_of_d = grad(self.distribution.calcForGradient)
    // Градиент считается приближенно, через величины функции на сетке --- взять соседние точки и решить линейное уравнение, например.

    return DistanceBetweenPoints(point1, point2) / gradient_modulus;
}

double StatsAPI::CalcIntegralOnFullIsoline(const IsolineCoords& isoline_points) const {
    double sum = 0;
    for (int i = 0; i < isoline_points.size()-1; ++i)
        sum += CalcIntegralOnInfinitesimalCurve(isoline_points[i], isoline_points[i+1]);

    return sum + CalcIntegralOnInfinitesimalCurve(isoline_points[isoline_points.size()-1], isoline_points[0]);
}

double StatsAPI::CalcIntegralOnLeftCurveOnIsoline(const IsolineCoords& isoline_points, int start_index, int end_index) const {
    if (start_index == end_index)
        throw std::invalid_argument("CalcIntegralOnLeftCurveOnIsoline: start_index equals end_index, start_index="+std::to_string(start_index));

    double sum = 0;

    if (end_index < start_index) {
        for (int i = end_index; i < start_index; ++i)
            sum += CalcIntegralOnInfinitesimalCurve(isoline_points[i], isoline_points[i+1]);
        return sum;
    }

    for (int i = 0; i < start_index; ++i)
        sum += CalcIntegralOnInfinitesimalCurve(isoline_points[i], isoline_points[i+1]);
    for (int i = end_index; i < isoline_points.size()-1; ++i)
        sum += CalcIntegralOnInfinitesimalCurve(isoline_points[i], isoline_points[i+1]);
    return sum + CalcIntegralOnInfinitesimalCurve(isoline_points[isoline_points.size()-1], isoline_points[0]);
}

double StatsAPI::CalcIntegralOnRightCurveOnIsoline(const IsolineCoords& isoline_points, int start_index, int end_index) const {
    if (start_index == end_index)
        throw std::invalid_argument("CalcIntegralOnRightCurveOnIsoline: start_index equals end_index, start_index="+std::to_string(start_index));

    double sum = 0;

    if (start_index < end_index) {
        for (int i = start_index; i < end_index; ++i)
            sum += CalcIntegralOnInfinitesimalCurve(isoline_points[i], isoline_points[i+1]);
        return sum;
    }

    for (int i = 0; i < end_index; ++i)
        sum += CalcIntegralOnInfinitesimalCurve(isoline_points[i], isoline_points[i+1]);
    for (int i = start_index; i < isoline_points.size()-1; ++i)
        sum += CalcIntegralOnInfinitesimalCurve(isoline_points[i], isoline_points[i+1]);
    return sum + CalcIntegralOnInfinitesimalCurve(isoline_points[isoline_points.size()-1], isoline_points[0]);
}

double StatsAPI::CalculateSStat(const CoordsOfPoint& point, double value) const {
    auto isoline_points = classic_grid->GetIsolineCoords(value);
    int point_index = InsertPointIntoIsoline(point, isoline_points);

    CoordsOfPoint zero_point = {0, 0, 0};
    auto mean = classic_grid->Func()->Mean();
    auto normed_mean = GetNormalizedVector(zero_point, mean);
    auto angles_of_mean = classic_grid->GetAnglesOfPoint(normed_mean); // TODO should I change distribution mean with the actual PDF's maximum?

    double mean_phi = angles_of_mean[0]; // azimuthal angle
    double mean_theta = angles_of_mean[1]; // polar angle

    if (normed_mean[0] == 0 and normed_mean[1] == 0) // if mean vector points to one of the poles, correct it's polar angle
        mean_theta = 0;

    // see my graduation work in order to understand the logic below

    std::vector<std::array<double, 2>> angles_of_isoline_points;
    angles_of_isoline_points.reserve(isoline_points.size());
    for (const auto& p : isoline_points)
        angles_of_isoline_points.push_back(classic_grid->GetAnglesOfPoint(p));
    
    std::vector<int> indices_between_zero_and_theta_mean;
    indices_between_zero_and_theta_mean.reserve(isoline_points.size()); // not optimal but at least something
    for (int i = 0; i < angles_of_isoline_points.size(); ++i)
        if (0 <= angles_of_isoline_points[i][1] && angles_of_isoline_points[i][1] <= mean_theta)
            indices_between_zero_and_theta_mean.push_back(i);


    std::vector<int> indices_between_zero_and_pi_minus_theta_mean;
    indices_between_zero_and_pi_minus_theta_mean.reserve(isoline_points.size()); // not optimal but at least something
    for (int i = 0; i < angles_of_isoline_points.size(); ++i)
        if (0 <= angles_of_isoline_points[i][1] && angles_of_isoline_points[i][1] <= (M_PI - mean_theta))
            indices_between_zero_and_pi_minus_theta_mean.push_back(i);

    std::vector<std::array<double, 2>> intersection_points;
    intersection_points.reserve(2);

    double smallest_diff_to_theta_mean = -1; // maybe should use intersection_points.size() for check instead

    // WARNING: all differencies below are calculated as |a-b|, since a and b must be non-negative angles!

    if (indices_between_zero_and_theta_mean.size() >= 2) { // TODO what if only 1 point?
        // find the smallest two differences to theta mean

        double first_smallest_diff = std::fabs(angles_of_isoline_points[indices_between_zero_and_theta_mean[0]][0] - angles_of_mean[0]);
        double second_smallest_diff = std::fabs(angles_of_isoline_points[indices_between_zero_and_theta_mean[1]][0] - angles_of_mean[0]);

        int index_of_first_smallest = 0; // in indices_between_zero_and_theta_mean vector
        int index_of_second_smallest = 1;

        if (first_smallest_diff > second_smallest_diff) {
            std::swap(first_smallest_diff, second_smallest_diff);
            std::swap(index_of_first_smallest, index_of_second_smallest);
        }

        for (int i = 2; i < indices_between_zero_and_theta_mean.size(); ++i) {
            double diff = std::fabs(angles_of_isoline_points[indices_between_zero_and_theta_mean[i]][0] - angles_of_mean[0]);
            if (diff < first_smallest_diff) {
                second_smallest_diff = first_smallest_diff;
                first_smallest_diff = diff;

                index_of_second_smallest = index_of_first_smallest;
                index_of_first_smallest = i;

            } else if (diff < second_smallest_diff) {
                second_smallest_diff = diff;
                index_of_second_smallest = i;
            }
        }

        smallest_diff_to_theta_mean = first_smallest_diff;

        intersection_points.push_back(angles_of_isoline_points[indices_between_zero_and_theta_mean[index_of_first_smallest]]);
        intersection_points.push_back(angles_of_isoline_points[indices_between_zero_and_theta_mean[index_of_second_smallest]]);
    }

    if (indices_between_zero_and_pi_minus_theta_mean.size() >= 2) { // TODO what if only 1 point?
        double theta_mean_plus_pi = std::fmod(angles_of_mean[0] + M_PI, 2 * M_PI);
        // find the smallest two differences to (theta mean + PI)

        double first_smallest_diff = std::fabs(angles_of_isoline_points[indices_between_zero_and_pi_minus_theta_mean[0]][0] - theta_mean_plus_pi);
        double second_smallest_diff = std::fabs(angles_of_isoline_points[indices_between_zero_and_pi_minus_theta_mean[1]][0] - theta_mean_plus_pi);

        int index_of_first_smallest = 0; // in indices_between_zero_and_pi_minus_theta_mean vector
        int index_of_second_smallest = 1;

        if (first_smallest_diff > second_smallest_diff) {
            std::swap(first_smallest_diff, second_smallest_diff);
            std::swap(index_of_first_smallest, index_of_second_smallest);
        }

        for (int i = 2; i < indices_between_zero_and_pi_minus_theta_mean.size(); ++i) {
            double diff = std::fabs(angles_of_isoline_points[indices_between_zero_and_pi_minus_theta_mean[i]][0] - theta_mean_plus_pi);
            if (diff < first_smallest_diff) {
                second_smallest_diff = first_smallest_diff;
                first_smallest_diff = diff;

                index_of_second_smallest = index_of_first_smallest;
                index_of_first_smallest = i;

            } else if (diff <= second_smallest_diff) {
                second_smallest_diff = diff;
                index_of_second_smallest = i;
            }
        }

        if (smallest_diff_to_theta_mean != -1) {
            if (smallest_diff_to_theta_mean > first_smallest_diff) {
                intersection_points.clear();
                intersection_points.push_back(angles_of_isoline_points[indices_between_zero_and_pi_minus_theta_mean[index_of_first_smallest]]);
                intersection_points.push_back(angles_of_isoline_points[indices_between_zero_and_pi_minus_theta_mean[index_of_second_smallest]]);
            }
        } else {
            intersection_points.push_back(angles_of_isoline_points[indices_between_zero_and_pi_minus_theta_mean[index_of_first_smallest]]);
            intersection_points.push_back(angles_of_isoline_points[indices_between_zero_and_pi_minus_theta_mean[index_of_second_smallest]]);
        }
    }

    std::array<double, 2> intersection_point = {(intersection_points[0][0] + intersection_points[1][0]) / 2, (intersection_points[0][1] + intersection_points[1][1]) / 2};
    auto intersection_point_coords = classic_grid->GetCoordsOfPoint(intersection_point);
    int intersection_point_index = InsertPointIntoIsoline(intersection_point_coords, isoline_points);

    auto point_A = isoline_points[GetIndexOfNextIsolinePoint(intersection_point_index, isoline_points)];
    auto vec_A = GetNormalizedVector(normed_mean, point_A);
    auto vec_I = GetNormalizedVector(normed_mean, intersection_point_coords);

    double isoline_integral = CalcIntegralOnFullIsoline(isoline_points);
    double curve_integral = 0;
    bool clockwiseDirection = false;
    if (CheckClockwiseDirection(normed_mean, vec_I, vec_A)) {
        clockwiseDirection = true;
        curve_integral = CalcIntegralOnRightCurveOnIsoline(isoline_points, intersection_point_index, point_index);
    } else {
        curve_integral = CalcIntegralOnLeftCurveOnIsoline(isoline_points, intersection_point_index, point_index);
    }

    // return clockwiseDirection too
    // return {"isoline": isoline_points, "point": intersection_point_coords, "S": curve_integral / isoline_integral}
    return curve_integral / isoline_integral;
}

double StatsAPI::CalculateTStat(double value) const {
    return spiral_grid->CalcIntegralInsideIsoline(value);
}

const std::vector<CoordsOfPoint> StatsAPI::Points(void) const {
    return points;
}

void StatsAPI::Validate(void) const {
    if (classic_grid == nullptr) {
        throw std::invalid_argument("invalid StatsAPI: classic_grid is nullptr");
    }
    if (spiral_grid == nullptr) {
        throw std::invalid_argument("invalid StatsAPI: spiral_grid is nullptr");
    }
    if (points.size() == 0) {
        throw std::invalid_argument("invalid StatsAPI: empty points");
    }
}

double StatsAPI::CalculateFunc(const CoordsOfPoint& point) const {
    return spiral_grid->Func()->Calc(point);
}
